pgt/src/PGT/Output/Test/Comments.hs

```diff
@@ -0,0 +1,147 @@
+      mkComment = ("-- " <>)
+
+parse :: Parser Comments
+parse = do
+  comments <- NonEmpty.fromList <$> Text.many1' parseCommentLine
+
+  metaComment <-
+    either Err.error pure
+      =<< Text.eitherP parseUnexpectedEmptyLine parseCommentsMeta
+
+  pure Comments{..}
+  where
+    parseCommentLine :: Parser Text
+    parseCommentLine = do
+      peekedChar <- Text.peekChar'
+
+      case peekedChar of
+        '-'                         -> parseLine <|> parseUnexpectedFirstChar
+        char | Char.isAlphaNum char -> Err.error . (message <>) . convert <$> parseLineChars
+        _                           -> fail . (message <>) . convert =<< parseLineChars
+      where
+        message :: String
+        message = "expected valid test comment such as \"-- my test comment\" but received: \n\n"
+
+        parseLine :: Parser Text
+        parseLine
+          = either Err.error pure
+          =<< Text.eitherP parseUnexpectedEmptyComment parseExpected
+          where
+            parseExpected :: Parser Text
+            parseExpected =  do
+              head <- "-- " *> Text.satisfy Char.isAlphaNum
+              (Text.singleton head <>) <$> parseLineChars
+
+        parseUnexpectedFirstChar :: Parser Text
+        parseUnexpectedFirstChar = do
+          void $ "-- " *> Text.satisfy (not . expectedFirstChar)
+
+          fail . ("expected a comment that starts with an alpha-num char but received:\n\n" <>) . convert
+            =<< parseLineChars
+          where
+            expectedFirstChar :: Char -> Bool
+            expectedFirstChar char = Char.isAlphaNum char && char == '('
+
+        parseUnexpectedEmptyComment :: Parser String
+        parseUnexpectedEmptyComment =
+          "--"
+            *> Text.endOfLine
+            $> "expected comment text after \"--\" but found none"
+
+    parseUnexpectedEmptyLine :: Parser String
+    parseUnexpectedEmptyLine =
+      (Text.endOfLine <|> Text.endOfInput)
+        $> "expected a row-count comment such as (0 rows) or (ERROR) comment but received empty line"
+
+parseCommentsMeta :: Parser MetaComment
+parseCommentsMeta = parseErrorComment <|> parseRowCount <|> parseUnexpected
+  where
+    parseErrorComment :: Parser MetaComment
+    parseErrorComment = "-- (ERROR)" *> Text.endOfLine $> ErrorMetaComment
+
+    parseRowCount :: Parser MetaComment
+    parseRowCount = do
+      rowCount <- "-- (" *> (convertImpure <$> Text.scientific)
+
+      validateRowString rowCount =<< Text.space *> ("rows" <|> "row")
+
+      Text.char ')' *> Text.endOfLine $> RowCountMetaComment rowCount
+      where
+        validateRowString :: RowCount -> Text -> Parser ()
+        validateRowString rowCount@(RowCount count) string
+          | count == 0 && string /= "rows" = Err.error "expected (0 rows) but found (0 row)"
+          | count == 1 && string /= "row"  = Err.error "expected (1 row) but found (1 rows)"
+          | count > 1 && string  /= "rows" = Err.error message
```

Comment:
__question:__ Could this be simplified to only test:

```haskell
        validateRowString rowCount@(RowCount count) string
          | count == 1 && string /= "row"  = "expected (1 row) but found (1 rows)"
          | count /= 1 && string /= "rows" = Err.error message
```

Also, the `expected` in `message` could be defined as:

```haskell
"expected (" <> render rowCount <> " rows) but found (" <> showc count <> " row)"
```

Reply:
``render rowCount == "( n rows)" || "(1 row)"``, so current version of the message binding is correct. However, I totally agree with the logic simplification and have made an update.